### PR TITLE
AlmaLinux 9, Oracle Linux 9, Red Hat Enterprise Linux 9

### PR DESCRIPTION
**Pull request by juliogonzalez**
# WARNING
Salt bundle did not work out of the box and some adjustments were needed. @vzhestkov took care of that, so please check with him and **do not merge** until his changes are ready for the bundle.

The following repository was used from Victor's PoC. Changes there should be hopefully soon get merged at `products`:
https://download.opensuse.org/repositories/home:/vzhestkov:/saltstack:/bundle:/next/AlmaLinux_9/

## What does this PR change?

Add AlmaLinux 9, Oracle Linux 9, Red Hat Enterprise Linux 9

## What's missing?

- Java changes
- PostgreSQL changes

## Tested

For reference: Alma Linux worked fine, as it seems it trusts it's GPG key. For Oracle Linux, if the instance was installed from scratch, and you never updated, you need to trust it before onboarding salt-ssh minions by installing a package, or calling `rpm --import`, or trust it after the onboarding with `rpm --import`. This is a known issue.

### AlmaLinux 9

- [x] Reposync, including errata
- [x] Bootstrap repo generation
- [x] CLM
- [x] Onboarding (bootstrap script, webui regular minion, webui salt ssh minion)
- [x] Install/remove packages/detect updates (regular minion and sallt ssh minion)
- [x] Salt remote command (regular minion and sallt ssh minion)
- [x] Enable locale formula and change the timezona, apply the highstate (regular minion and sallt ssh minion)

### Oracle Linux 9

- [x] Reposync, including errata
- [x] Bootstrap repo generation
- [x] CLM
- [x] Onboarding (bootstrap script, webui regular minion, webui salt ssh minion)
- [x] Install/remove packages/detect updates (regular minion and sallt ssh minion)
- [x] Salt remote command (regular minion and sallt ssh minion)
- [x] Enable locale formula and change the timezona, apply the highstate (regular minion and sallt ssh minion)

### RHEL9

Nothing

## GUI diff

No difference.

- [x] **DONE**

## Documentation
- Documentation issue was created: Pending

- [ ] **DONE**

## Test coverage
- Cucumber tests to be added later

- [x] **DONE**

## Links

- [x] **DONE**

## Changelogs

Make sure the changelogs entries you are adding are compliant with https://github.com/uyuni-project/uyuni/wiki/Contributing#changelogs and https://github.com/uyuni-project/uyuni/wiki/Contributing#uyuni-projectuyuni-repository

If you don't need a changelog check, please mark this checkbox:

- [ ] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)

## Re-run a test

If you need to re-run a test, please mark the related checkbox, it will be unchecked automatically once it has re-run:

- [ ] Re-run test "changelog_test"
- [ ] Re-run test "backend_unittests_pgsql"
- [ ] Re-run test "java_pgsql_tests"
- [ ] Re-run test "schema_migration_test_pgsql"
- [ ] Re-run test "susemanager_unittests"
- [ ] Re-run test "javascript_lint"
- [ ] Re-run test "spacecmd_unittests"

### CODE REVIEW

**Review by mcalmer — APPROVED** [review]

**Comment by vzhestkov** [comment]
Just a small note that there is no yet subproject of Alma9 in the main bundle project, and in testing it's not yet building, will create SR for testing to fix the building for alma9

**Comment by juliogonzalez** [comment]
> Just a small note that there is no yet subproject of Alma9 in the main bundle project, and in testing it's not yet building, will create SR for testing to fix the building for alma9

That's right, but I'd suggest merging now, so we don't have to backport to a future Manager-4.3 branch :-)

**Comment by juliogonzalez** [comment]
Let's see if I can add Oracle and RHEL during the Learning Tuesday, to leave things ready.

File: utils/spacewalk-common-channels.ini [hunk 1]
```diff
@@ -1469,6 +1469,73 @@ gpgkey_id = %(_uyuni_gpgkey_id)s
 gpgkey_fingerprint = %(_uyuni_gpgkey_fingerprint)s
 repo_url = https://download.opensuse.org/repositories/systemsmanagement:/Uyuni:/Master:/SLE15-Uyuni-Client-Tools/SLE_15/
 
+[oraclelinux9]
+archs    = x86_64, aarch64
+name     = Oracle Linux 9 (%(arch)s)
+checksum = sha256
+gpgkey_url =
```
Comment (juliogonzalez) [inline utils/spacewalk-common-channels.ini:1476]: The doc [mentions](https://linux.oracle.com/security/gpg/) https://yum.oracle.com/RPM-GPG-KEY-oracle-ol9, but it returns 404, so I didn't add it.

Be aware, however that the GPG "file" contains in fact *two keys*:

```
[root@m211 ~]# gpg --show-keys /etc/pki/rpm-gpg/RPM-GPG-KEY-oracle 
gpg: directory '/root/.gnupg' created
gpg: keybox '/root/.gnupg/pubring.kbx' created
pub   rsa4096 2022-01-19 [SC] [expires: 2042-01-14]
      3E6D826D3FBAB389C2F38E34BC4D06A08D8B756F
uid                      Oracle Linux (release key 1) <secalert_us@oracle.com>
sub   rsa4096 2022-01-19 [E] [expires: 2041-06-01]

pub   rsa4096 2022-01-19 [SC] [expires: 2042-01-14]
      982231759C7467065D0CE9B2A7DD07088B4EFBE6
uid                      Oracle Linux (backup key 1) <secalert_us@oracle.com>
sub   rsa4096 2022-01-19 [E] [expires: 2041-06-02]
```
AFAIK `spacewalk-common-channels` does *not* support two keys, so if they ever change to the backup key, we'll need to change it here, and at the DB with a SQL script.

**Comment by juliogonzalez** [comment]
Requesting re-review, as I added Oracle 9, and the bootstrap repo for RHEL9 (untested)

File: utils/spacewalk-common-channels.ini (hunk 1 above)
Comment (mcalmer) [reply utils/spacewalk-common-channels.ini:1476]: There is no oraclelinux:9 docker image yet. But guessing with the help of v8 we should try

`file:///etc/pki/rpm-gpg/RPM-GPG-KEY-oracle`

as URL. This is what we will do for all EL clones and point to the local installed GPG key.

Comment (juliogonzalez) [reply utils/spacewalk-common-channels.ini:1476]: Still, you are not telling me to use this at `spacewalk-common-channels`, right? `file:///etc/pki/rpm-gpg/RPM-GPG-KEY-oracle` does not exist at the Server, even if I added it to the build-keys... or?

Comment (juliogonzalez) [reply utils/spacewalk-common-channels.ini:1476]: No, it does not:
```
m225:~ # ls -l /etc/pki/rpm-gpg/RPM-GPG-KEY-oracle
ls: cannot access '/etc/pki/rpm-gpg/RPM-GPG-KEY-oracle': No such file or directory

File: utils/spacewalk-common-channels.ini [hunk 2]
```diff
@@ -1469,6 +1469,73 @@ gpgkey_id = %(_uyuni_gpgkey_id)s
...
+archs    = x86_64
+name     = Latest RDMA packages for Unbreakable Enterprise Kernel on Oracle Linux 9
+base_channels = oraclelinux9-%(arch)s
+repo_url = https://yum.oracle.com/repo/OracleLinux/OL9/RDMA/%(arch)s/
+
+[oraclelinux9-uyuni-client]
+name     = Uyuni Client Tools for %(base_channel_name)s
+archs    = %(_x86_archs)s, aarch64
+base_channels = oraclelinux9-%(arch)s
+gpgkey_url = https://download.opensuse.org/repositories/systemsmanagement:/Uyuni:/Stable:/EL9-Uyuni-Client-Tools/EL_9/repodata/repomd.xml.key
```
Comment (mcalmer) [inline utils/spacewalk-common-channels.ini:1525]: We may need to change this when we are ready with the GPG key feature.

File: utils/spacewalk-common-channels.ini (hunk 1 above)
Comment (juliogonzalez) [reply utils/spacewalk-common-channels.ini:1476]: Clarified at slack, and changed.

File: utils/spacewalk-common-channels.ini [hunk 3]
```diff
@@ -3155,6 +3222,113 @@ gpgkey_id = 2F86D6A1
...
+gpgkey_id = %(_uyuni_gpgkey_id)s
+gpgkey_fingerprint = %(_uyuni_gpgkey_fingerprint)s
+repo_url = https://download.opensuse.org/repositories/systemsmanagement:/Uyuni:/Master:/EL9-Uyuni-Client-Tools/EL_9/
+
+[almalinux9-epel9]
+label    = epel9-%(base_channel)s
+name     = EPEL 9 for %(base_channel_name)s
+archs    = x86_64, aarch64, ppc64le, s390x
+base_channels = almalinux9-%(arch)s
+gpgkey_url = http://download.fedoraproject.org/pub/epel/RPM-GPG-KEY-EPEL-9
```
Comment (juliogonzalez) [inline utils/spacewalk-common-channels.ini:3327]: I did *not* change it here, as it's this key is not available at the clients by default, saame as the client tools, so I guess you will handle it later.

**Comment by juliogonzalez** [comment]
Ready to be reviewed again. I hope it's OK this time.

BTW @mcalmer... I guess we want to merge the PRs for Ubuntu 22.04 and openEuler 22.03 ASAP?

File: utils/spacewalk-common-channels.ini (hunk 1 above)
Comment (mcalmer) [reply utils/spacewalk-common-channels.ini:1476]: Just to write it down: this is the path on the client. The file is installed by oracle

File: susemanager/susemanager.changes [hunk 4]
```diff
@@ -1,3 +1,6 @@
+- Add Oracle Linux 9 bootstrap repositories for Uyuni
+- Add AlmaLinux 9 bootstrap repositories for Uyuni
+- Add Red Hat Enterprise Linux 9 repositories for Uyuni
```
Comment (mcalmer) [inline susemanager/susemanager.changes:3]: "Red Hat Enterprise Linux 9" ? We do not have RES yet, nor do we provide URLs for native RHEL9.
Is it a good idea to provide it already?

**Review by mcalmer — APPROVED** [review]

File: susemanager/src/mgr_bootstrap_data.py [hunk 5]
```diff
@@ -1497,6 +1513,22 @@
         'BASECHANNEL' : 'almalinux8-aarch64', 'PKGLIST' : RES8,
         'DEST' : DOCUMENT_ROOT + '/pub/repositories/almalinux/8/bootstrap/'
     },
+    'almalinux-9-x96_64-uyuni' : {
```
Comment (santeri3700) [inline susemanager/src/mgr_bootstrap_data.py:1516]: Looks like there is a typo in the AL9 architecture part, `x96_64` should be `x86_64`. Not sure if this really affects anything, but I guess it's worth a mention.

https://github.com/uyuni-project/uyuni/blob/b69b5582a4dffc83e69cf53e6058a1bb4bf19d08/susemanager/src/mgr_bootstrap_data.py#L1516

**Pull request merged**